Series 2 i2c driver enhancements

fimohame opened the pull request:
Updated the I2C driver for EFR Series 2 devices based on the design improvements in the GSDK I2C implementation

fimohame committed:
drivers: i2c: silabs: Update i2c driver with HAL changes

Updated the I2C driver for EFR Series 2 devices based on the
design improvements in the SiSDK I2C implementation

Signed-off-by: S Mohamed Fiaz <Fiaz.Mohamed@silabs.com>
Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -506,7 +529,7 @@ static DEVICE_API(i2c, i2c_silabs_dev_driver_api) = {
 	static const struct i2c_silabs_dev_config i2c_silabs_dev_config_##idx = {                  \
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),                                       \
-		.base = (I2C_TypeDef *)DT_INST_REG_ADDR(idx),                                      \
+		.peripheral = SL_PERIPHERAL_I2C##idx,                                              \
```
This doesn't work, the devicetree index doesn't need to match the peripheral instance number. You need to construct a `sl_peripheral_t` from the `reg` and `clocks` data from devicetree. We cannot use the predefined ones from SiSDK in Zephyr.

fimohame replied:
I've addressed the comment

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -503,10 +526,17 @@ static DEVICE_API(i2c, i2c_silabs_dev_driver_api) = {
+	static const sl_peripheral_val_t i2c_peripheral_val_##idx = {                              \
+		.base = DT_INST_REG_ADDR(idx),                                                     \
+		.clk_branch = DT_INST_CLOCKS_CELL(idx, branch),                                    \
+		.bus_clock = (DT_INST_CLOCKS_CELL(idx, enable) ? &i2c_bus_clock_##idx : NULL),     \
```
peripheral struct wants a pointer to the bus clock, not the bus clock itself.

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -64,15 +61,15 @@ static int i2c_silabs_pm_action(const struct device *dev, enum pm_device_action
-#ifdef CONFIG_I2C_SILABS_DMA
 	struct i2c_silabs_dev_data *data = dev->data;
 
+	if (!IS_ENABLED(CONFIG_I2C_SILABS_DMA)) {
```
addressed

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -227,24 +230,46 @@ static int i2c_silabs_transfer_sync(const struct device *dev, struct i2c_msg *ms
+							msgs[i].len,
+							CONFIG_I2C_SILABS_TIMEOUT) != 0) {
+				k_sem_give(&data->bus_lock);
+				return -ETIMEDOUT;
```
addressed

fimohame committed:
drivers: i2c: silabs: remove dev arg from PM state lock helpers

Dropped the device argument from i2c_silabs_pm_policy_state_lock_get/put.
Updated all internal usage.

Signed-off-by: S Mohamed Fiaz <Fiaz.Mohamed@silabs.com>
Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -167,95 +151,146 @@ static int i2c_silabs_transfer_dma(const struct device *dev, struct i2c_msg *msg
+				if (sl_i2c_follower_receive_non_blocking(&data->i2c_handle,
+									 msgs[i].buf, msgs[i].len,
+									 NULL) != 0) {
+					k_sem_give(&data->bus_lock);
```
as this is an async transfer, using pattern of goto wouldn't be ideal in this api

asmellby reviewed (changes requested):
The first 3 commits need to be squashed, otherwise the PR breaks bisectability.
* Updating the HAL without also updating the implementation will cause compile errors
* Follower mode support is not new, and shouldn't be introduced as a "new feature" separate commit after first being silently removed in the previous commit. The HAL just split the functions handling the two cases.
